### DISCUSSION TITLE
[2021/11/09] Frontend Night

### DISCUSSION THREAD

**Discussion by jjenko** [wantedly/frontend-night#176] [General]
## Policy

- 資料の準備をがんばらない
- 参加・途中退席は自由
- スキップしない

話したい・聞きたいネタを書いてく ✏️

ハッシュタグ [#frontend_night](https://twitter.com/hashtag/frontend_night?src=hash)

Prev: https://github.com/wantedly/frontend-night/issues/165 

 _Created from https://github.com/wantedly/frontend-night/issues/65 by [issue-creator](https://github.com/rerost/issue-creator)_

**Comment by kobayang** [comment]
## [React Freeze](https://github.com/software-mansion-labs/react-freeze)

Suspence のメカニズムを利用して、ユーザーに見えていない範囲のコンポーネントを unmount するのではなく、state を保持させたまま凍結させる。

unmount しないからスクロール位置、入力ステート、ロードされた画像が変更されずに保持されるようになる。

```tsx
function SomeComponent({ shouldSuspendRendering }) {
  return (
    <Freeze freeze={shouldSuspendRendering}>
      <MyOtherComponent />
    </Freeze>
  );
}
```

https://blog.swmansion.com/experimenting-with-react-freeze-71da578e2fa6

**Reply by kobayang** [reply]
Freeze コンポーネントの実装
https://github.com/software-mansion-labs/react-freeze/blob/main/src/index.tsx

`Suspender` の中身

```tsx
const promiseCache = useRef<StorageRef>({}).current;
// 1. freeze 時の描画で promise を作り、 throw する
if (freeze && !promiseCache.promise) {
  promiseCache.promise = new Promise((resolve) => {
    promiseCache.resolve = resolve;
  });
  throw promiseCache.promise;
// 2. freeze中は promise を throw し続ける
} else if (freeze) {
  throw promiseCache.promise;
// 3. promise がある場合は、resolve し、 children を描画。
} else if (promiseCache.promise) {
  promiseCache.resolve!();
  promiseCache.promise = undefined;
}
// それ以外の場合も描画

return <Fragment>{children}</Fragment>;
```

freeze されたタイミングで、 promiseCache.promise を作り、それを保持し throw し続ける。
freeze が解除されたタイミングで、children を render する。

throw した promise は Suspense が キャッチする。

**Reply by kobayang** [reply]
ここのおもしろいポイントは、 promise を throw し続けることで、unmount を防いでいること。

**Reply by kobayang** [reply]
メインのユースケースは React Navigation の stack view とあるが、freeze の 切り替えをいい感じに実装できれば web の方でも、Virtualization のユースケースで利用できそう。

**Reply by kobayang** [reply]
> Virtualization のユースケースで利用できそう。

作ってみた
https://github.com/kobayang/react-freeze-v12n
